freebsd 12 FUSE xattr issues

The xattrs in the native freebsd filesystem are set like this, then backed up with borg:
```
   xattr.setxattr(os.path.join(self.input_path, 'fusexattr'), 'user.foo', b'bar')
   xattr.setxattr(os.path.join(self.input_path, 'fusexattr'), 'user.empty', b'')
```

Then the archive is fuse mounted and test_fuse executed:

freebsd 12 amd64 VM (see Vagrantfile), current master branch, no fakeroot:

```
freebsd64: >           assert sorted(no_selinux(xattr.listxattr(out_fn))) == ['user.empty', 'user.foo', ]
freebsd64: E           AssertionError: assert ['empty', 'foo'] == ['user.empty', 'user.foo']
freebsd64: E             At index 0 diff: 'empty' != 'user.empty'
```

borg 1.2.0.dev569+g23bd2345.d20180704

----
freebsd 12 amd64 VM (see Vagrantfile), current platform-xattr branch, no fakeroot:

borg 1.2.0.dev560+gadb298d1

behaves the same after setting `borg.xattr.XATTR_FAKEROOT = False` so that same test code is executed.

----

So it seems that something is stripping away the "user." from the xattr keys.

freebsd 12 amd64 VM (see Vagrantfile), current platform-xattr branch, no fakeroot:

When having XATTR_FAKEROOT = True, this happens:
```
                if not xattr.XATTR_FAKEROOT and xattr.is_enabled(self.input_path):
                    assert sorted(no_selinux(xattr.listxattr(out_fn))) == ['user.empty', 'user.foo', ]
                    assert xattr.getxattr(out_fn, 'user.foo') == b'bar'
                    # Special case: getxattr returns None (not b'') when reading an empty xattr.
                    assert xattr.getxattr(out_fn, 'user.empty') is None
                else:
>                   assert xattr.listxattr(out_fn) == []

src/borg/testsuite/archiver.py:2192: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/borg/platform/freebsd.pyx:62: in borg.platform.freebsd.listxattr
    n, buf = _listxattr_inner(func, path)
src/borg/platform/xattr.py:72: in _listxattr_inner
    n = _check(func(path, buf, size), path, detect_buffer_too_small=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

rv = -1, path = b'/tmp/tmpyrlor7jv/mountpoint/input/fusexattr', detect_buffer_too_small = True

    def _check(rv, path=None, detect_buffer_too_small=False):
        if rv < 0:
            e = get_errno()
            if detect_buffer_too_small and e == errno.ERANGE:
                # listxattr and getxattr signal with ERANGE that they need a bigger result buffer.
                # setxattr signals this way that e.g. a xattr key name is too long / inacceptable.
                raise BufferTooSmallError
            else:
                try:
                    msg = os.strerror(e)
                except ValueError:
                    msg = ''
                if isinstance(path, int):
                    path = '<FD %d>' % path
>               raise OSError(e, msg, path)
E               OSError: [Errno 22] Invalid argument: b'/tmp/tmpyrlor7jv/mountpoint/input/fusexattr'

src/borg/platform/xattr.py:55: OSError
```
So listxattr on a file that is visible via the fuse mount and has not xattrs set gives errno 22.

--- a/src/borg/testsuite/xattr.py
+++ b/src/borg/testsuite/xattr.py
@@ -44,13 +44,13 @@ class XattrTestCase(BaseTestCase):
     def test_listxattr_buffer_growth(self):
         # make it work even with ext4, which imposes rather low limits
         buffer.resize(size=64, init=True)
-        # xattr raw key list will be size 9 * (10 + 1), which is > 64
-        keys = ['user.attr%d' % i for i in range(9)]
+        # xattr raw key list will be > 64
+        keys = ['user.attr%d' % i for i in range(20)]
         for key in keys:
             setxattr(self.tmpfile.name, key, b'x')
         got_keys = listxattr(self.tmpfile.name)
         self.assert_equal_se(got_keys, keys)
-        self.assert_equal(len(buffer), 128)
+        self.assert_true(len(buffer) > 64)
 
     def test_getxattr_buffer_growth(self):
         # make it work even with ext4, which imposes rather low limits

--- a/src/borg/xattr.py
+++ b/src/borg/xattr.py
@@ -323,6 +323,7 @@ elif sys.platform.startswith('freebsd'):  # pragma: freebsd only
     libc.extattr_set_file.argtypes = (c_char_p, c_int, c_char_p, c_char_p, c_size_t)
     libc.extattr_set_file.restype = c_int
     ns = EXTATTR_NAMESPACE_USER = 0x0001
+    prefix = 'user.'
 
     def listxattr(path, *, follow_symlinks=True):
         def func(path, buf, size):
@@ -335,7 +336,7 @@ elif sys.platform.startswith('freebsd'):  # pragma: freebsd only
                     return libc.extattr_list_link(path, ns, buf, size)
 
         n, buf = _listxattr_inner(func, path)
-        return [os.fsdecode(name) for name in split_lstring(buf[:n]) if name]
+        return [prefix + os.fsdecode(name) for name in split_lstring(buf[:n]) if name]
 
     def getxattr(path, name, *, follow_symlinks=True):
         def func(path, name, buf, size):
@@ -347,6 +348,10 @@ elif sys.platform.startswith('freebsd'):  # pragma: freebsd only
                 else:
                     return libc.extattr_get_link(path, ns, name, buf, size)
 
+        # strip namespace if there, but ignore if not there.
+        # older borg / attic versions did not prefix the namespace to the names.
+        if name.startswith(prefix):
+            name = name[len(prefix):]
         n, buf = _getxattr_inner(func, path, name)
         return buf[:n] or None
 
@@ -360,6 +365,10 @@ elif sys.platform.startswith('freebsd'):  # pragma: freebsd only
                 else:
                     return libc.extattr_set_link(path, ns, name, value, size)
 
+        # strip namespace if there, but ignore if not there.
+        # older borg / attic versions did not prefix the namespace to the names.
+        if name.startswith(prefix):
+            name = name[len(prefix):]
         _setxattr_inner(func, path, name, value)
 
 else:  # pragma: unknown platform only

